### PR TITLE
Add SBOM and related checks to project

### LINKED ISSUE #139: Add software bill of materials (SBOM) to the project

### PR DESCRIPTION
**Pull request by d33bs**
# Description

This PR adds a software bill of materials (SBOM) to the project using the SPDX convention. It also adds a pre-commit based check to ensure the `packages` section of the SBOM is up to date with any changes which have occurred in the source which could be reflected as a difference in the SBOM (meaning the SBOM will stay up to date so long as checks are passing).

SBOM's are increasingly becoming commonplace to include within software projects and the use is also listed in RSMM framework (https://arxiv.org/pdf/2406.01788) under focus area 4 software adoptability, section 4.6 deployability as "the project has an SBOM".

This change is a first step towards following the convention and then checking for this convention in other work through the Almanack.

Closes #139 

## What is the nature of your change?

- [ ] Content additions or updates (adds or updates content)
- [ ] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).
- [ ] Breaking change (these changes would cause existing functionality to not work as expected).

# Checklist

Please ensure that all boxes are checked before indicating that this pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own contributions.
- [x] I have commented my content, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to related documentation (outside of book content).
- [x] My changes generate no new warnings.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests that prove my additions are effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

**Review by gwaybio — COMMENTED** [review]
Is there a place to describe what these checks actually are? (sorry if i missed this)

**Comment by d33bs** [comment]
Thanks @gwaybio - I added some comments to help better describe the check for the SBOM. All changes here pertain to the Almanack repo itself; we're adding an `sbom.json` file and a pre-commit check which makes sure this file is up to date. The pre-commit check uses a poethepoet shell script to ensure the diff is inconsequential (the date and checksum will differ slightly from file to file based on how they're generated but we don't need to update if the dependencies haven't changed). We use `syft` to help generate the SPDX-style SBOM. We use a newly added dev dependency `sbomdiff` to help with this diffing and to avoid more complicated solutions.

Eventually my hope would be to check for these files in other repos and make this a metric for the Almanack. The changes here were complex enough that I felt they should be standalone before an external-facing metric would be developed.

File: pyproject.toml [hunk 1]
```diff
@@ -134,6 +135,33 @@ if [ $(echo "$output" | wc -l) -gt 1 ]; then
...
+# If an existing SBOM file doesn't exist, save the new one as the baseline
+if [ ! -f sbom.json ]; then
+  echo "No existing SBOM found. Creating baseline SBOM."
+  mv sbom-new.json sbom.json
+  exit 1
+fi
+
+# Use sbomdiff to check for changes
+if ! sbomdiff sbom-new.json sbom.json; then
+  echo "SBOM has changed! Please review and commit the updated SBOM."
```
Comment (falquaddoomi) [inline pyproject.toml:156]: This is probably a silly question, but how should someone review the SBOM? From what I can see it's not exactly a human-readable file; perhaps including the `sbomdiff` output in this message, or at least hinting that they might use that tool, might help them review it?

**Review by falquaddoomi — APPROVED** [review]
Neat; I wasn't aware that there were formal formats for SBOMs, so thanks for the opportunity to learn. I left a minor comment, but otherwise this all looks reasonable to me.

File: pyproject.toml (hunk 1 above)
Comment (d33bs) [reply pyproject.toml:156]: Thanks good question - the SBOM I feel can be thought of as project data / metadata, similar to that of a lockfile (which isn't usually reviewed line by line). In my testing of this I found that the `sbomdiff` output does display from pre-commit runs if and where there are issues (it shows a table of where differences were detected). The check here works similarly to that of Black or ruff-format - the first run shows a failure and automatically fixes the issue (notifying the developer that something changed). The second run passes as the file has been amended for commit into the git history.

**Comment by d33bs** [comment]
Thanks @falquaddoomi and @gwaybio for the reviews! Merging this in.

**Pull request merged**